rules/cache.go

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package rules
+
+import "github.com/m3db/m3metrics/rules"
+
+// Source is a datasource providing match results.
+type Source interface {
+	// IsValid returns whether a given version is still valid.
+	IsValid(version int) bool
+
+	// Match returns the match result for an given id.
+	Match(id []byte) rules.MatchResult
+}
+
+// Cache caches the rule matching result associated with metrics.
+type Cache interface {
+	// Get returns the rule matching result associated with a metric id.
+	Get(namespace string, id []byte) rules.MatchResult
+
+	// Set sets the result source for a given namespace.
+	Set(namespace string, source Source)
```

Comment:
Would it not be slightly more idiomatic and matching the APIs of source to have the cache interface be more specific that `Get(...)` actually calls `Match(...)` and separate set/delete naming with something more like:
```go
type Cache interface {
  Match(namespace string, id []byte) rules.MatchResult
  RegisterSource(namespace string, source Source)
  UnregisterSource(namespace string)
  // ...
}
```